src/mono/mono/mini/interp/transform.c

```diff
@@ -8364,49 +8373,141 @@ interp_mark_reachable_bblocks (TransformData *td)
 	}
 }
 
+/**
+ * Returns TRUE if in_bb defines a variable that is used in
+ * the block containing @cond_ins, FALSE otherwise.
+ */
+
 static gboolean
-interp_prev_ins_defines_var (InterpInst *ins, int var1, int var2)
+interp_prev_block_defines_var (InterpInst *in_bb_ins, InterpInst *cond_ins)
 {
+	InterpInst *prev_ins = interp_prev_ins(cond_ins);
+	int var1 = -1, var2 = -1;
+	if (MINT_IS_BINOP_CONDITIONAL_BRANCH(cond_ins->opcode)) {
+		var1 = cond_ins->sregs [0];
+		var2 = cond_ins->sregs [1];
+	} else {
+		if (prev_ins && MINT_IS_COMPARE(prev_ins->opcode)) {
+			var1 = prev_ins->sregs [0];
+			var2 = prev_ins->sregs [1];
+		} else {
+			var1 = cond_ins->sregs [0];
+		}
+	}
+
 	// Check max of 5 instructions
+	prev_ins = in_bb_ins;
 	for (int i = 0; i < 5; i++) {
-		ins = interp_prev_ins (ins);
-		if (!ins)
+		prev_ins = interp_prev_ins (prev_ins);
+		if (!prev_ins)
 			return FALSE;
-		if (mono_interp_op_dregs [ins->opcode] && (ins->dreg == var1 || ins->dreg == var2))
+		if (mono_interp_op_dregs [prev_ins->opcode] && (prev_ins->dreg == var1 || prev_ins->dreg == var2))
 			return TRUE;
 	}
 	return FALSE;
 }
 
+/**
+ * Check if the given basic block has a known pattern for inlining into callers blocks, if so, return a pointer to the conditional branch instruction.
+ *
+ * The known patterns are:
+ * - `branch`: a branch instruction (either MINT_IS_CONDITIONAL_BRANCH or MINT_BR).
+ * - `ldc; branch`: a load instruction followed by a binary conditional branch.
+ * - `compare; branch`: a compare instruction followed by a unary conditional branch.
+ * - `ldc; compare; branch`: a load instruction followed by a compare instruction and a unary conditional branch.
+ */
+static gboolean
+interp_inline_into_callers(InterpBasicBlock *bb, InterpInst** cond_ins) {
+	InterpInst *first = interp_first_ins (bb);
+	InterpInst *second = first ? interp_next_ins(first) : NULL;
+	InterpInst *third = second ? interp_next_ins(second) : NULL;
+	
```

Comment:
Maybe compiler is smart enough. But you could avoid checking for non null with every single if like:
```
if (!first) return;
...
checking patterns with first instruction
...
if (!second) return;
...
checking patterns with first and second instructions
...
```

Reply:
Good point, updated.